vertx-auth-mtls/src/main/java/io/vertx/ext/auth/authentication/CertificateCredentials.java

```diff
@@ -0,0 +1,114 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.vertx.ext.auth.authentication;
+
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.Collections;
+import java.util.List;
+
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * Certificates based {@link Credentials} implementation, carries user's certificates, which can be used for
+ * authenticating or authorizing users.
+ */
+public class CertificateCredentials implements Credentials
+{
+    private final List<Certificate> certificateChain;
+    private final X509Certificate peerCertificate;
+
+    public CertificateCredentials(Certificate certificate)
+    {
+        this(Collections.singletonList(certificate));
+    }
+
+    public CertificateCredentials(List<Certificate> certificateChain)
+    {
+        this.certificateChain = Collections.unmodifiableList(certificateChain);
+        this.peerCertificate = getPeerCertificate();
+    }
+
+    /**
+     * Create {@link CertificateCredentials} from {@link HttpServerRequest}
+     * @return CertificateCredentials
+     */
+    public static CertificateCredentials fromHttpRequest(HttpServerRequest request)
+    {
+        try
+        {
+            return new CertificateCredentials(request.connection().peerCertificates());
+        }
+        catch (Exception e)
+        {
+            throw new IllegalArgumentException("Could not extract certificates from request", e);
+        }
+    }
+
+    /**
+     * @return The certificate chain contained in {@link CertificateCredentials}
+     */
+    public List<Certificate> certificateChain()
+    {
+        return certificateChain;
+    }
+
+    /**
+     * @return peer's certificate. It does not return null value once {@link #checkValid()} passes
+     */
+    public X509Certificate peerCertificate()
+    {
+        return peerCertificate;
+    }
+
+    public void checkValid() throws CredentialValidationException
+    {
+        checkValid(this);
+    }
+
+    @Override
+    public <V> void checkValid(V arg) throws CredentialValidationException
+    {
+        if (certificateChain.isEmpty())
+        {
+            throw new CredentialValidationException("Certificate Chain cannot be empty");
+        }
+    }
+
+    /**
+     * @deprecated {@link CertificateCredentials} currently not represented in Json format.
+     */
+    @Override
+    public JsonObject toJson()
+    {
+        throw new UnsupportedOperationException("Deprecated authentication method");
+    }
+
+    private X509Certificate getPeerCertificate()
+    {
+        // First certificate in the chain is peer's own cert
+        if (!certificateChain.isEmpty() && certificateChain.get(0) instanceof X509Certificate)
+        {
+            return (X509Certificate) certificateChain.get(0);
+        }
+
+        return null;
```

Comment:
Should we allow for the first X509Certificate in the chain?
```suggestion
                // First certificate in the chain is peer's own cert
        return certificateChain.stream()
                               .filter(cert -> cert instanceof X509Certificate)
                               .map(cert -> (X509Certificate) cert)
                               .findFirst()
                               .orElse(null);
```

Reply:
Since the first entry is usually the private key entry, we are checking for that

Reply:
> Should we allow for the first X509Certificate in the chain?

No. The behavior is get the first cert and proceed if it is a X509Certificate